### ISSUE TITLE
No gradients provided for any variable, check your graph for ops that do not support gradients, between variables. Problem with tf.cast op

### ISSUE THREAD

**Issue by alejandroaguileraalcalde-ing** [tensorflow/tensorflow#47083]
I am creating a model using tf v1 and I got this problem and I dont know how to solve it, any ideas? I think is due to tf.cast op, but i need to use it so if a number is < 1, it turns into 0

Error: No gradients provided for any variable, check your graph for ops that do not support gradients, between variables

My code:

#Ejemplo para 4 datos: Location, Email, IMEI y Device ID: 
filtracion_location = tf.placeholder(tf.float32, name='location_input')
filtracion_email = tf.placeholder(tf.float32, name='email_input')
filtracion_imei = tf.placeholder(tf.float32, name='imei_input')
filtracion_device = tf.placeholder(tf.float32, name='device_input')

#objetivo: answer: 1 ò 0
y_ = tf.placeholder(tf.float32, name='target')

#distintos pesos para cada dato: nombre= Px
Plocation = tf.Variable(9., name='Plocation')
Pemail = tf.Variable(8., name='Pemail')
Pimei = tf.Variable(3., name='Pimei')
Pdevice = tf.Variable(2., name='Pdevice')

#umbral de decisión
umbral = tf.Variable(10., name='umbral')

#variables auxiliares: 
y_aux1 = tf.Variable(0.)
y_aux2 = tf.Variable(0.)
y_aux3 = tf.Variable(0.)
y_aux4 = tf.Variable(0.)
y_aux5 = tf.Variable(0.)
y_aux6 = tf.Variable(0.)
y_aux7 = tf.Variable(0.)

#salida del modelo
y_aux1 = tf.add(tf.multiply(filtracion_location, Plocation), 0.0)
y_aux2 = tf.add(tf.multiply(filtracion_email, Pemail), y_aux1)
y_aux3 = tf.add(tf.multiply(filtracion_imei, Pimei), y_aux2)
y_aux4 = tf.add(tf.multiply(filtracion_device, Pdevice), y_aux3)

#conseguir que la salida sea 0 ò 1
y_aux5 = tf.divide(y_aux4, umbral)
y_aux6 = tf.cast(y_aux5, tf.int16)
y_aux7 = tf.cast(y_aux6, tf.float32)

y = tf.div_no_nan(y_aux7, y_aux7)
y = tf.identity(y, name='output')

loss = tf.reduce_mean(tf.square(y - y_))
optimizer = tf.train.GradientDescentOptimizer(learning_rate=0.01)
train_op = optimizer.minimize(loss, name='train')  ---> here is the error

init = tf.global_variables_initializer()

# Creating a tf.train.Saver adds operations to the graph to save and
# restore variables from checkpoints.

saver_def = tf.train.Saver().as_saver_def()

with open('graph_Ant.pb', 'wb') as f:
  f.write(tf.get_default_graph().as_graph_def().SerializeToString())

**Comment by alejandroaguileraalcalde-ing** [comment]
I also tried with tf.round operation instead of tf.cast and there was the same problem

**Comment by amahendrakar** [comment]
@alejandroaguileraalcalde-ing,
TensorFlow 1.x is not actively supported. Could you please update TensorFlow to the latest stable version v2.4.1 and check if you are facing the same issue. Thanks!

**Comment by alejandroaguileraalcalde-ing** [comment]
I am using TF 1.x becouse I am creating a app that works with "training on device" for a Federated Machine Learning project, so I need the checkpoints and graph file that TF v1 creates when using save method. 

Could you tell me the list of operations that support gradients? I have search for it and didnt get any result.

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with TF v1.15.5. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/aeeba6d12f3bd6ed916574271790045a/47083.ipynb). Thanks!

**Comment by alejandroaguileraalcalde-ing** [comment]
ok, but  could you show me or tell me where can I find the list of operations that suppport gradientes?

**Comment by saxenasaurabh** [comment]
I think the cast to `int16` is the problem here. That would cause `Cast` to [drop](https://github.com/tensorflow/tensorflow/blob/ab438b386b5b95ed310fd9c31a9c5570a9fc6db3/tensorflow/python/ops/math_grad.py#L1960) gradients.
